tiledb/sm/array_schema/attribute.cc

```diff
@@ -303,22 +290,36 @@ void Attribute::serialize(
 }
 
 void Attribute::set_cell_val_num(unsigned int cell_val_num) {
-  if (type_ == Datatype::ANY) {
+  check_cell_val_num(cell_val_num);
+  cell_val_num_ = cell_val_num;
+  set_default_fill_value();
+}
+
+void Attribute::check_cell_val_num(unsigned int cell_val_num) const {
+  if (type_ == Datatype::ANY && cell_val_num != constants::var_num) {
     throw AttributeStatusException(
         "Cannot set number of values per cell; Attribute datatype `ANY` is "
         "always variable-sized");
   }
 
-  if (order_ != DataOrder::UNORDERED_DATA && type_ != Datatype::STRING_ASCII &&
-      cell_val_num != 1) {
-    throw AttributeStatusException(
-        "Cannot set number of values per cell; An ordered attribute with "
-        "datatype '" +
-        datatype_str(type_) + "' can only have cell_val_num=1.");
+  // If ordered, check the number of values of cells is supported.
+  if (order_ != DataOrder::UNORDERED_DATA) {
+    if (type_ == Datatype::STRING_ASCII) {
+      if (cell_val_num != constants::var_num) {
+        throw std::invalid_argument(
```

Comment:
`invalid_argument` is the wrong exception here. This should be `AttributeException`, currently called `AttributeStatusException`. As a transition tactic to keep the scope of the PR down, go ahead and change the name of the exception class and add an alias to the old name.

The principle here is that we use `std` exceptions to catch implementation defects. We use `tdb` exceptions for faults that can be caused by bad user input or generally usage outside of documented limits.

It's worth noting that `set_cell_val_num` is only called (outside of test code) by the C API, where there's an API call of similar name. That's directly a user error. The other constructors are used for deserialization either from storage or capnp; these are not user errors, although there are security concerns around malformed data. In situations such as this we favor the user-facing exception type.